From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mykyta Komarn <nkomarn@hotmail.com>
Date: Wed, 30 Sep 2020 18:17:51 -0700
Subject: [PATCH] Heavily optimize recipe lookups

Recipe lookups are now cached in CraftingManager. I also rewrote the stream replacement code to not create an ArrayList, rather just use the Collection that was already there. Additionally, an EMPTY_MAP variable was added to prevent bottlenecks during map creation, since that map is only ever iterated.

NonNullList now uses GlueList internally as its delegate list, which should bring general improvements to all tile entities with an inventory.

These changes knock off ~40ms of tick duration with a sample of ~7,700 running furnaces on a server.

diff --git a/src/main/java/net/minecraft/server/CraftingManager.java b/src/main/java/net/minecraft/server/CraftingManager.java
index de7d5ed8d6260ce5ee4164df29cdaf69f561045c..22b8166ebb8e407d49bb73cf09163a11d31d1666 100644
--- a/src/main/java/net/minecraft/server/CraftingManager.java
+++ b/src/main/java/net/minecraft/server/CraftingManager.java
@@ -25,6 +25,10 @@ public class CraftingManager extends ResourceDataJson {
     private static final Logger LOGGER = LogManager.getLogger();
     public Map<Recipes<?>, Object2ObjectLinkedOpenHashMap<MinecraftKey, IRecipe<?>>> recipes = ImmutableMap.of(); // CraftBukkit
     private boolean d;
+    // Yatopia start
+    public static final Object2ObjectLinkedOpenHashMap<MinecraftKey, IRecipe<?>> EMPTY_MAP = new Object2ObjectLinkedOpenHashMap<>();
+    public static final Map<Recipes<?>, List> CACHE = new Object2ObjectLinkedOpenHashMap<>();
+    // Yatopia end
 
     public CraftingManager() {
         super(CraftingManager.a, "recipes");
@@ -79,8 +83,8 @@ public class CraftingManager extends ResourceDataJson {
     public <C extends IInventory, T extends IRecipe<C>> Optional<T> craft(Recipes<T> recipes, C c0, World world) {
         // CraftBukkit start
         // Yatopia start - replace stream
-        List<IRecipe<C>> collection = new java.util.ArrayList<>(this.b(recipes).values());
-        Optional<T> recipe = collection.isEmpty() ? Optional.empty() : recipes.a(collection.get(0), world, c0);
+        Collection<IRecipe<C>> collection = this.b(recipes).values(); // Yatopia
+        Optional<T> recipe = collection.isEmpty() ? Optional.empty() : recipes.a(collection.iterator().next(), world, c0); // Yatopia
         /*
         Optional<T> recipe = this.b(recipes).values().stream().flatMap((irecipe) -> {
             return SystemUtils.a(recipes.a(irecipe, world, c0));
@@ -108,19 +112,25 @@ public class CraftingManager extends ResourceDataJson {
     }
 
     public <C extends IInventory, T extends IRecipe<C>> List<T> a(Recipes<T> recipes) {
+        // Yatopia Start - Replaced Logic
+        return CACHE.computeIfAbsent(recipes, recipes1 -> {
+            return new net.yatopia.server.list.GlueList<>(this.b(recipes).values());
+        }); // Yatopia
+        /* 
         List<IRecipe<C>> list = new ArrayList<>();
         for (IRecipe<C> irecipe : this.b(recipes).values()) {
             IRecipe<C> ciRecipe = irecipe;
             list.add(ciRecipe);
         }
         return (List) list;
+        */ // Yatopia End
     }
 
     public <C extends IInventory, T extends IRecipe<C>> List<T> b(Recipes<T> recipes, C c0, World world) {
         // Yatopia start - WHY?! WHO DID THIS TO YOU!?!
         Collection<IRecipe<C>> recipeCollection = this.b(recipes).values();
         if (recipeCollection.isEmpty()) return java.util.Collections.emptyList();
-        List<T> ret = new java.util.ArrayList<>();
+        List<T> ret = new net.yatopia.server.list.GlueList<>(); // Yatopia
         for (IRecipe<C> iRecipe : recipeCollection) recipes.a(iRecipe, world, c0).ifPresent(ret::add);
         ret.sort(Comparator.comparing((iRecipe) -> iRecipe.getResult().getTranslationKey()));
         return ret;
@@ -134,7 +144,7 @@ public class CraftingManager extends ResourceDataJson {
     }
 
     private <C extends IInventory, T extends IRecipe<C>> Map<MinecraftKey, IRecipe<C>> b(Recipes<T> recipes) {
-        return (Map) this.recipes.getOrDefault(recipes, new Object2ObjectLinkedOpenHashMap<>()); // CraftBukkit
+        return (Map) this.recipes.getOrDefault(recipes, EMPTY_MAP); // CraftBukkit // Yatopia
     }
 
     public <C extends IInventory, T extends IRecipe<C>> NonNullList<ItemStack> c(Recipes<T> recipes, C c0, World world) {
diff --git a/src/main/java/net/minecraft/server/NonNullList.java b/src/main/java/net/minecraft/server/NonNullList.java
index 2d3c406a05b9710d40471aa61c3540dbc971c1e3..f8fe6d5133de9fd10f18813027a36fc232eb6caf 100644
--- a/src/main/java/net/minecraft/server/NonNullList.java
+++ b/src/main/java/net/minecraft/server/NonNullList.java
@@ -10,7 +10,7 @@ import org.apache.commons.lang3.Validate;
 
 public class NonNullList<E> extends AbstractList<E> {
 
-    private final List<E> a;
+    private final net.yatopia.server.list.GlueList<E> a; // Yatopia
     private final E b;
 
     public static <E> NonNullList<E> a() {
@@ -22,7 +22,7 @@ public class NonNullList<E> extends AbstractList<E> {
         Object[] aobject = new Object[i];
 
         Arrays.fill(aobject, e0);
-        return new NonNullList<>(Arrays.asList(aobject), e0);
+        return new NonNullList<>((List<E>) Arrays.asList(aobject), e0); // Yatopia - decompile fix
     }
 
     @SafeVarargs
@@ -31,11 +31,11 @@ public class NonNullList<E> extends AbstractList<E> {
     }
 
     protected NonNullList() {
-        this(Lists.newArrayList(), (Object) null);
+        this(Lists.newArrayList(), null); // Yatopia - decompile fix
     }
 
     protected NonNullList(List<E> list, @Nullable E e0) {
-        this.a = list;
+        this.a = new net.yatopia.server.list.GlueList<>(list); // Yatopia
         this.b = e0;
     }
 
diff --git a/src/main/java/net/minecraft/server/RecipeItemStack.java b/src/main/java/net/minecraft/server/RecipeItemStack.java
index 30da7471c3ecc66a61cb9fe1dd58d6d65438c505..27978f7b52c8db02f69fdcb092ffcce4550904d3 100644
--- a/src/main/java/net/minecraft/server/RecipeItemStack.java
+++ b/src/main/java/net/minecraft/server/RecipeItemStack.java
@@ -32,7 +32,7 @@ public final class RecipeItemStack implements Predicate<ItemStack> {
 
     public void buildChoices() {
         if (this.choices == null) {
-            List<ItemStack> list = new ArrayList<>();
+            List<ItemStack> list = new net.yatopia.server.list.GlueList<>(); // Yatopia
             Set<ItemStack> uniqueValues = new HashSet<>();
             for (Provider recipeitemstack_provider : this.b) {
                 for (ItemStack itemStack : recipeitemstack_provider.a()) {
